packages/studio-base/src/components/ForwardContextProviders.tsx

```diff
@@ -0,0 +1,68 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { useContext, useLayoutEffect, useState } from "react";
+import { StoreApi, createStore, useStore } from "zustand";
+
+import { useMustNotChange, useShallowMemo } from "@foxglove/hooks";
+import MultiProvider from "@foxglove/studio-base/components/MultiProvider";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type ForwardedContexts = ReadonlyMap<React.Context<any>, StoreApi<{ value: unknown }>>;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type ForwardedContextValues = ReadonlyMap<React.Context<any>, unknown>;
+
+/**
+ * Returns a Map entry for forwarding the given context's value, which can be passed to
+ * `ForwardContextProviders`.
+ */
+export function useForwardContext<T>(
+  context: React.Context<T>,
+): readonly [React.Context<T>, StoreApi<{ value: T }>] {
+  const value = useContext(context);
+  const [store] = useState(() => createStore(() => ({ value })));
+  useLayoutEffect(() => {
+    store.setState({ value });
+  }, [store, value]);
+  return useShallowMemo([context, store] as const);
+}
+
+function getContextValues(contexts: ForwardedContexts): ForwardedContextValues {
```

Comment:
```suggestion
function getContextValueMap(contexts: ForwardedContexts): ForwardedContextValues {
```
maybe also update `ForwardedContextValues`? feels a bit confusing since it's  a map but sounds like an array